chain/src/pipe.rs

```diff
@@ -236,17 +244,115 @@ fn check_header_known(bh: Hash, ctx: &mut BlockContext) -> Result<(), Error> {
 
 /// Quick in-memory check to fast-reject any block we've already handled
 /// recently. Keeps duplicates from the network in check.
-fn check_known(bh: Hash, ctx: &mut BlockContext) -> Result<(), Error> {
+fn check_known(header: &BlockHeader, ctx: &mut BlockContext) -> Result<(), Error> {
+	let bh = header.hash();
 	if bh == ctx.head.last_block_h || bh == ctx.head.prev_block_h {
-		return Err(ErrorKind::Unfit("already known".to_string()).into());
+		return Err(ErrorKind::Unfit("already known in head".to_string()).into());
 	}
+
 	let cache = ctx.block_hashes_cache.read().unwrap();
 	if cache.contains(&bh) {
 		return Err(ErrorKind::Unfit("already known in cache".to_string()).into());
 	}
-	if ctx.orphans.contains(&bh) {
-		return Err(ErrorKind::Unfit("already known in orphans".to_string()).into());
+	Ok(())
+}
+
+fn check_known_orphans(header: &BlockHeader, ctx: &mut BlockContext) -> Result<(), Error> {
+	if ctx.orphans.contains(&header.hash()) {
+		Err(ErrorKind::Unfit("already known in orphans".to_string()).into())
+	} else {
+		Ok(())
+	}
+}
+
+// Check if this block is in the store already.
+fn check_known_store(header: &BlockHeader, ctx: &mut BlockContext) -> Result<(), Error> {
```

Comment:
> check_known_store() # slower (but still fast) lookup against full block index (2,000 recent blocks on fast sync node)

Where is this `2000` ?

Reply:
Its the threshold for triggering compaction in `check_compact()` -
https://github.com/mimblewimble/grin/blob/86aea875de1c4c6595fa1e1403048cea8134c71f/servers/src/common/adapters.rs#L545-L547